Filter properties in the database instead of in memory

alimi opened the pull request:
This PR has a few changes, but they were all motivated by changing how `RailsProperties::Base#properties` works. `#properties` accepts `var` as an argument and returns the target's matching property object. It has done this by fetching all of the targets property objects and filtering them in memory to find the requested `var`.

https://github.com/1debit/rails-properties/blob/e97fe12249e065c9d8f93c0185f441262c55d94b/lib/rails-properties/base.rb#L18

Although this filtering returns only one property object, the others are still loaded in memory and associated with the target. This can lead to strange side affects because callers aren't expecting the other objects to be loaded in memory. To fix this we can filter property objects at the database level so only the requested objects are loaded in memory.

```ruby
property_objects.find_or_initialize_by(var: var)
```

The [migration template](https://github.com/1debit/rails-properties/blob/e97fe12249e065c9d8f93c0185f441262c55d94b/lib/generators/rails_properties/migration/templates/migration.rb#L15) has an index on `target_type`, `target_id`, `var` so the database query should be performant.

To avoid conflicts when writing new property objects, memoize the requested property object after it's been initialized. This was done implicitly before when all the objects were loaded in memory. Without this memoization, test like [this](https://github.com/1debit/rails-properties/blob/e97fe12249e065c9d8f93c0185f441262c55d94b/spec/properties_spec.rb#L48-L50) fail.

Along with this change, we drop support for old versions of Rails. As a result, we can also remove `RailsProperties#can_protect_attributes?` and the related code. `#can_protect_attributes?` was used to support the Protected Attributes gem on old versions of Rails, and [Protected Attributes doesn't support Rails 5+](https://github.com/rails/protected_attributes/blob/a1e6dfd39cc106939a2f96ec98cfff4950eb3e7e/protected_attributes.gemspec#L21).

Commit:
Pin sqlite3 version for dev/test

  * Starting in Rails 6, ActiveRecord's Sqlite adapter restricted the
    sqlite3 gem version to ~> 1.4. This is still true in Rails 7.1. See
    https://github.com/rails/rails/pull/35844.
  * Use the same restriction here to ensure a compatible sqlite3 version
    is installed for the dev and test.
  * The version restriction has been relaxed on Rails edge but it hasn't
    landed in a released Rails version yet. See
    https://github.com/rails/rails/pull/51592.

Commit:
Drop support for older ActiveRecord versions

Commit:
Remove can_protect_attributes?

  * can_protect_attributes? was used to support old Rails versions that
    were using the Protected Attributes gem to migrate to strong
    parameters. Protected Attributes hasn't been supported since Rails
    5.0, and since rails-properties only supports 6+ we don't need
    can_protect_attributes? any more.

Commit:
Filter properties in the db instead of in memory

  * properties was loading all of a target's property objects into
    memory and filtering the collection to only return the var that was
    requested. Although the filtering returns only one property object,
    the others are still loaded in memory and associated with the
    target. This can lead to strange side affects because callers
    aren't expecting the other objects to be loaded in memory.
  * Instead of filtering the property objects in memory, filter them at
    the database level so only the requested objects are loaded in
    memory.
  * To avoid conflicts when writing new property objects, memoize the
    requested property object after it's been initialized. This was done
    implicitly before when all the objects were loaded in memory.
  * Update the queries_spec to reflect the new behavior—each request for
    a target's property objects will result in unique queries because
    filtering is done in the database and not in memory.

Commit:
Bump the major version

alimi commented on lib/rails-properties/base.rb:
```diff
@@ -12,11 +12,7 @@ def properties(var)
-          else
-            property_objects.detect { |s| s.var == var.to_s } || property_objects.build(:var => var.to_s, :target => self)
-          end
+          cached_properties[var] ||= property_objects.find_or_initialize_by(var: var)
```
This memoization is being done to support new property objects. If a property object isn't in the database, `property_objects.find_or_initialize_by(var: var)` returns an new object in memory. Since we can write multiple values to a new property like

```ruby
account.properties(:portal).premium = true
account.properties(:portal).fee = 42.5
```

`#properties` needs to return the same instance of the property object. Otherwise, we'll have two different property objects for the same `var` (`:portal` in the above example), and [that's not allowed](https://github.com/1debit/rails-properties/blob/e97fe12249e065c9d8f93c0185f441262c55d94b/lib/generators/rails_properties/migration/templates/migration.rb#L15).

This memoization was happening implicitly before when everything was being loading into memory.

alimi commented on lib/rails-properties/version.rb:
```diff
@@ -1,3 +1,3 @@
 module RailsProperties
-  VERSION = '3.4.3'
+  VERSION = '4.0.0'
```
This feels like a big enough change to warrant a major version bump.

alimi commented on spec/queries_spec.rb:
```diff
@@ -51,7 +51,7 @@
         user.properties(:calendar).bar = 'string'
         user.save!
-      }.to perform_queries(3)
+      }.to perform_queries(4)
```
The number of queries increases because we're not loading all of the target's property objects into memory anymore. So we have one query to get the user's `:dashboard` property object and another to get the user's `:calendar` property object.

jmmastey-chime reviewed (approved):
Is it reasonable to add a test that exercises the buggy behavior that we'd seen? Or is that a lot to set up?

alimi commented:
It'd be a lot of setup because a lot of different factors at play. But the underlying change that will fix the bug is covered by the `queries_spec` referenced in https://github.com/1debit/rails-properties/pull/5/files/e441b9a25d3d6c5b93de5d08f3c624ccd412d36a#r1620940022.

Commit:
Add failing spec that tests reload before

  * Calling reload on the target is not reseting the associated property
    objects anymore. This is broken because
    RailsProperties::Base#properties is memoizing the ActiveRecord
    property_objects records. This before was working before because
    ActiveRecord was doing the memoiziation for us. This should be fixed
    before merging.

alimi commented:
I just pushed up a failing spec in fe28690b1cec5da1c32183147462f015be8ee48e. Calling reload on the target is not resetting the associated property objects anymore. This is broken because `RailsProperties::Base#properties` is memoizing the ActiveRecord `property_objects` records. This was working before because ActiveRecord was doing the memoiziation for us. This should be fixed before merging; I saw this issue when testing these changes in a Rails app.

alimi commented:
Closing this since we haven't been able to figure out fixing the reload behavior.
